### PR TITLE
Remove support old when rules

### PR DESCRIPTION
**Pull request by ONS-Guilhem-Forey**
### PR Context
We want to get rid of the deprecated WHEN rules.
They include:
- Routing conditions
- Skip conditions
- Content variants
- Question variants

### How to review
Ticket is made of two PRs, validator and runner.

We need to merge the validator first, otherwise the runner's CI won't pass.

### Checklist

- [x] Upgrade all schemas to the new WHEN rules
- [x] Remove code related to deprecated when rules
- [x] Remove deprecated tests

### CODE REVIEW

**Comment by ONS-Guilhem-Forey** [comment]
[Remove support for old when rules (3 SP)](https://trello.com/c/uCidXQd2/5295-remove-support-for-old-when-rules-3-sp)

**Review by berroar — COMMENTED** [review]
If you temporarily update the validator branch in `scripts/run_validator.sh` so that it points at your branch rather than latest it will help show that the validator changes you've made are correct. Just need to remember to revert it before merging 👍

**Review by berroar — APPROVED** [review]
Checked through the changes and tested,  LGTM 👍

File: schemas/test/en/test_calculated_summary_dependent_questions.json [hunk 1]
```diff
@@ -139,9 +139,19 @@
                             "type": "CalculatedSummary",
                             "title": "We have calculated your total spend to be <em>%(total)s</em>. Is this correct?",
                             "calculation": {
-                                "calculation_type": "sum",
-                                "answers_to_calculate": ["answer-1", "answer-3"],
-                                "title": "Total capital expenditure"
+                                "operation": {
```
Comment (berroar) [inline schemas/test/en/test_calculated_summary_dependent_questions.json:142]: Not sure if the calculated summary schemas and tests are supposed to be updated as part of this? We me still be supporting the old calcualated summary format

File: app/questionnaire/path_finder.py [hunk 2]
```diff
@@ -286,14 +258,9 @@ def _remove_current_blocks_answers_for_backwards_routing(
...
             if "when" in rules:
-                if isinstance(rules["when"], dict):
-                    self._remove_current_blocks_answers_for_new_backwards_routing(
-                        rules["when"], answer_ids_for_current_block
-                    )
-                else:
-                    for rule in rules["when"]:
-                        if "id" in rule and rule["id"] in answer_ids_for_current_block:
-                            self.answer_store.remove_answer(rule["id"])
+                self._remove_current_blocks_answers_for_new_backwards_routing(
```
Comment (petechd) [inline app/questionnaire/path_finder.py:261]: Not "new" anymore since we remove old?

Comment (ONS-Guilhem-Forey) [reply app/questionnaire/path_finder.py:261]: Renamed it to `_remove_block_anwers_for_backward_routing_according_to_when_rule`
Kept the method as `"when" in rule` might still be False, e.g if the routing block doesn't have a when rule

File: schemas/test/en/test_calculated_summary_dependent_questions.json (hunk 1 above)
Comment (ONS-Guilhem-Forey) [reply schemas/test/en/test_calculated_summary_dependent_questions.json:142]: Thanks, reverted this

File: scripts/run_validator.sh [hunk 3]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env bash
 tag=latest
-TAG=${tag} docker-compose -f docker-compose-schema-validator.yml up -d
+TAG=5295-remove-support-for-old-when-rules docker-compose -f docker-compose-schema-validator.yml up -d
```
Comment (ONS-Guilhem-Forey) [inline scripts/run_validator.sh:3]: ⚠️ ⚠️ 

Note to self: revert this before merging 

⚠️ ⚠️

File: schemas/test/en/test_calculated_summary.json [hunk 4]
```diff
@@ -339,8 +381,18 @@
                             "id": "percentage-total-playback",
                             "title": "We calculate the total of percentage values entered to be %(total)s. Is this correct?",
                             "calculation": {
-                                "calculation_type": "sum",
-                                "answers_to_calculate": ["fifth-percent-answer", "sixth-percent-answer"],
+                                "operation": {
```
Comment (berroar) [inline schemas/test/en/test_calculated_summary.json:384]: Not sure I was quite clear in the last comment I made, but the removal of the old when rules is separate to calculated summary. For now we still need to support calculated summary in both the old and new format, so some of these schemas and tests will need to be reverted.

File: tests/integration/questionnaire/test_questionnaire_calculated_summary.py [hunk 5]
```diff
@@ -61,24 +61,8 @@ def test_calculated_summary_no_skip(self):
...
-        self.launchSurvey("test_new_calculated_summary")
-        self._complete_calculated_summary_path_no_skip()
-        self.assertNotInBody("Skipped Fourth")
-        self.assertInBody(
-            "We calculate the total of currency values entered to be £180.00"
-        )
-
-    def test_new_calculated_summary_repeating_section(self):
-        self.launchSurvey("test_new_calculated_summary_repeating_section")
+    def test_calculated_summary_repeating_section(self):
```
Comment (berroar) [inline tests/integration/questionnaire/test_questionnaire_calculated_summary.py:64]: The changes to these tests will need to be reverted

File: tests/functional/spec/features/calculated_summary.spec.js [hunk 6]
```diff
@@ -4,7 +4,4 @@ describe("Feature: Calculated Summary", () => {
   describe("Given I have a Calculated Summary", () => {
```
Comment (berroar) [inline tests/functional/spec/features/calculated_summary.spec.js:4]: Here as well we still need to test the old and new calculated formats for the time being 👍

File: tests/functional/spec/features/calculated_summary_repeating_section.spec.js [hunk 7]
```diff
@@ -1,33 +1,33 @@
...
+import PrimaryPersonListCollectorPage from "../../generated_pages/calculated_summary_repeating_section/primary-person-list-collector.page";
+import PrimaryPersonListCollectorAddPage from "../../generated_pages/calculated_summary_repeating_section/primary-person-list-collector-add.page.js";
+import ListCollectorPage from "../../generated_pages/calculated_summary_repeating_section/list-collector.page";
+import ListCollectorAddPage from "../../generated_pages/calculated_summary_repeating_section/list-collector-add.page";
 
 describe("Feature: Calculated Summary Repeating Section", () => {
   describe("Given I have a Calculated Summary in a Repeating Section", () => {
     before("Get to Calculated Summary", () => {
-      browser.openQuestionnaire("test_new_calculated_summary_repeating_section.json");
+      browser.openQuestionnaire("test_calculated_summary_repeating_section.json");
```
Comment (berroar) [inline tests/functional/spec/features/calculated_summary_repeating_section.spec.js:30]: As above, the same applies for the repeats

**Review by berroar — CHANGES_REQUESTED** [review]

File: tests/functional/spec/features/calculated_summary_repeating_section.spec.js (hunk 7 above)
Comment (ONS-Guilhem-Forey) [reply tests/functional/spec/features/calculated_summary_repeating_section.spec.js]: Reverted this test file after bringing back the schema

File: tests/functional/spec/features/calculated_summary.spec.js (hunk 6 above)
Comment (ONS-Guilhem-Forey) [reply tests/functional/spec/features/calculated_summary.spec.js]: Reverted this test file after bringing back the schema

File: tests/integration/questionnaire/test_questionnaire_calculated_summary.py (hunk 5 above)
Comment (ONS-Guilhem-Forey) [reply tests/integration/questionnaire/test_questionnaire_calculated_summary.py]: Reverted this test file after bringing back the schema

File: schemas/test/en/test_calculated_summary.json (hunk 4 above)
Comment (ONS-Guilhem-Forey) [reply schemas/test/en/test_calculated_summary.json]: Brought back the old calculated summary schemas whilst updating their when rules

File: tests/integration/test_app_create.py [hunk 8]
```diff
@@ -124,6 +124,7 @@ def test_csp_policy_headers(self):
             "EQ_ENABLE_LIVE_RELOAD": False,
             "CDN_URL": cdn_url,
             "ADDRESS_LOOKUP_API_URL": address_lookup_api_url,
+            "EQ_APPLICATION_VERSION": "v1.0.0",
```
Comment (MebinAbraham) [inline tests/integration/test_app_create.py:127]: Unrelated change?, Will need a `application-version` file as per README.

File: app/questionnaire/path_finder.py [hunk 9]
```diff
@@ -196,19 +195,7 @@ def _evaluate_routing_rules(
...
-                    self.schema,
-                    self.metadata,
-                    self.answer_store,
-                    self.list_store,
-                    current_location=this_location,
-                    routing_path_block_ids=routing_path_block_ids,
-                )
-            else:
-                should_goto = should_goto_new(rule, when_rule_evaluator)
+            should_goto = check_should_goto(rule, when_rule_evaluator)
```
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply app/questionnaire/path_finder.py:198]: Applied, and renamed `should_goto` to `rule_valid`

File: schemas/test/en/test_variants_question.json [hunk 10]
```diff
@@ -4,9 +4,9 @@
     "schema_version": "0.0.1",
     "data_version": "0.0.3",
     "survey_id": "0",
-    "title": "Test Question Variants",
+    "title": "Test New Question Variants",
```
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply schemas/test/en/test_variants_question.json:7]: Removed

File: tests/app/data_model/test_list_store.py [hunk 11]
```diff
@@ -151,3 +151,13 @@ def test_first_raises_index_error_when_list_is_empty():
...
+
+
+def test_get_item_using_method():
+    store = ListStore()
+
+    first_id = store.add_list_item("people")
+
+    item = store.get("people")
+
+    assert item.items[0] == first_id
```
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply tests/app/data_model/test_list_store.py:163]: As discussed in Slack, is is because old rules resolved it a diff way.
I've tested without this test, and the coverage fails.

File: tests/integration/test_app_create.py (hunk 8 above)
Comment (ONS-Guilhem-Forey) [reply tests/integration/test_app_create.py:127]: Removed, and added `.application-version`, which kept the coverage to 100%.

File: tests/app/questionnaire/test_path_finder.py [hunk 12]
```diff
@@ -260,10 +223,7 @@ def test_routing_path_with_conditional_value_not_in_metadata(answer_store, list_
 
 @pytest.mark.parametrize(
     "schema, expected_routing_path_ids",
-    (
-        ("test_new_skip_condition_block", ["do-you-want-to-skip"]),
-        ("test_skip_condition_block", ["do-you-want-to-skip", "a-non-skipped-block"]),
-    ),
+    (("test_skip_condition_block", ["do-you-want-to-skip"]),),
 )
 def test_new_routing_path_should_skip_block(
```
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply tests/app/questionnaire/test_path_finder.py:228]: Unnecessary occurences of `parametrize` removed. `new` removed from method name.

File: tests/app/forms/test_questionnaire_form.py [hunk 13]
```diff
@@ -1449,16 +1448,15 @@ def test_generate_form_with_title_and_no_answer_label(app, answer_store, list_st
...
-            form = generate_form(
-                schema,
-                question_schema,
-                answer_store,
-                list_store,
-                metadata=get_metadata(),
-                response_metadata={},
-                form_data=form_data,
-            )
+        form = generate_form(
```
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply tests/app/forms/test_questionnaire_form.py:1451]: Guessing because there is no routing rule nor when rule in the schema that test is using. That is, `test_title.json`

File: schemas/test/en/test_routing_date_not_equals.json [hunk 14]
```diff
@@ -44,32 +44,29 @@
                                         "id": "single-date-answer",
                                         "label": "Today",
                                         "mandatory": true,
-                                        "type": "Date"
+                                        "type": "MonthYearDate"
```
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply schemas/test/en/test_routing_date_not_equals.json:47]: This when rule references it in this format: https://github.com/ONSdigital/eq-questionnaire-runner/blob/d37637ca90ee50e17d6227559178cbf40d46aab9/schemas/test/en/test_routing_date_not_equals.json#L63

File: schemas/test/en/test_calculated_summary_repeating_section.json [hunk 15]
```diff
@@ -0,0 +1,887 @@
+{
```
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply schemas/test/en/test_calculated_summary_repeating_section.json:1]: Deleted the file. We should have only `test_new_calculated_summary_repeating_section.json`, as on `main`.
The `new` stands for new calculated summary, not new when rules.

File: schemas/test/en/test_new_calculated_summary_repeating_section.json [hunk 16]
```diff
@@ -1,887 +1,887 @@
...
-        },
-        {
-            "name": "ru_name",
-            "type": "string"
-        }
-    ],
-    "questionnaire_flow": {
-        "type": "Hub",
-        "options": {}
+  "mime_type": "application/json/ons/eq",
```
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply schemas/test/en/test_new_calculated_summary_repeating_section.json]: Mistake. Fixed it, this file has no change now, which reduces the diff size.

File: schemas/test/en/test_routing_date_not_equals.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply schemas/test/en/test_routing_date_not_equals.json:47]: Ah I see, but that will still convert to a full date, it will essentially become `2018-02-01`. Happy to keep 👍

File: app/questionnaire/variants.py [hunk 17]
```diff
@@ -20,26 +19,16 @@ def choose_variant(
     for variant in block.get(variants_key, []):
         when_rules = variant.get("when", [])
```
Comment (MebinAbraham) [inline app/questionnaire/variants.py:20]: "when" for variants are mandatory.
```suggestion
        when_rules = variant["when"]
```

**Review by MebinAbraham — APPROVED** [review]
👍

**Review by berroar — APPROVED** [review]

File: tests/app/forms/test_questionnaire_form.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ONS-Guilhem-Forey) [reply tests/app/forms/test_questionnaire_form.py:1451]: It seems to me that it was always redundant

File: app/questionnaire/variants.py (hunk 17 above)
Comment (ONS-Guilhem-Forey) [reply app/questionnaire/variants.py]: Applied

**Review by petechd — APPROVED** [review]
